File: bumperbot_python_examples/bumperbot_python_examples/simple_qos_publisher.py
import rclpy
from rclpy.node import Node 
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy
from std_msgs.msg import String

class SimpleQoSPublisher(Node):
    def __init__(self):
        super().__init__("simple_qos_publisher")

        self.qos_profile_ = QoSProfile(depth=10)

        self.declare_parameter("reliability", "system_default")
        self.declare_parameter("durability", "system_default")

        reliability = self.get_parameter("reliability").get_parameter_value().string_value
        durability = self.get_parameter("durability").get_parameter_value().string_value

        if reliability == "best_effort":
            self.qos_profile_.reliability = QoSReliabilityPolicy.BEST_EFFORT
            self.get_logger().info("[Reliability] : Best Effort")
        elif reliability == "reliable":
            self.qos_profile_.reliability = QoSReliabilityPolicy.RELIABLE
            self.get_logger().info("[Reliability] : Reliable")
        elif reliability == "system_default":
            self.qos_profile_.reliability = QoSReliabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info("[Reliability] : System Default")
        else:
            self.get_logger().error(f"Selected Reliability QoS: {reliability} doesn't exists")
            return

        if durability == "transient_local":
            self.qos_profile_.durability = QoSDurabilityPolicy.TRANSIENT_LOCAL
            self.get_logger().info("[Durability] : Transient Local")
        elif durability == "volatile":
            self.qos_profile_.durability = QoSDurabilityPolicy.VOLATILE
            self.get_logger().info("[Durability] : Volatile")
        elif durability == "system_default":
            self.qos_profile_.reliability = QoSDurabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info("[Durability] : System Default")
        else:
            self.get_logger().error(f"Selected Reliability QoS: {durability} doesn't exists")
            return

        self.pub_ = self.create_publisher(String, "chatter", qos_profile=self.qos_profile_)
        self.counter_ = 0
        self.frequency = 1.0

        self.get_logger().info(f"Publishing at {self.frequency} HZ")
        self.timer_ = self.create_timer(1/self.frequency, self.timerCallback)

    def timerCallback(self):
        self.counter_ += 1
        msg = String()
        msg.data = f"timer: {self.counter_}"
        self.pub_.publish(msg)
        


def main(args=None):
    rclpy.init(args=args)
    node = SimpleQoSPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
    

File: bumperbot_cpp_examples/src/simple_qos_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <chrono>

using namespace std::chrono_literals;

class SimpleQoSPublisher: public rclcpp::Node
{
    public:
        SimpleQoSPublisher(): Node("simple_publisher"), qos_profile_pub_(10), counter_(0)
        {
            declare_parameter<std::string>("reliability", "system_default");
            declare_parameter<std::string>("durability", "system_default");

            const auto reliability = get_parameter("reliability").as_string();
            const auto durability = get_parameter("durability").as_string();

            if(reliability == "best_effort"){
                qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);
                RCLCPP_INFO(get_logger(), "[Reliability]: Best Effort");
            }else if(reliability == "reliable"){
                qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);
                RCLCPP_INFO(get_logger(), "[Reliability]: Reliable");
            }else if(reliability == "system_default"){
                qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT);
                RCLCPP_INFO(get_logger(), "[Reliability]: System Default");
            }else{
                RCLCPP_ERROR(get_logger(), "Selected Reliability QoS doesn't exists");
            }

            if(durability == "transient_local"){
                qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
                RCLCPP_INFO(get_logger(), "[Reliability]: transient local");
            }else if(durability == "volatile"){
                qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
                RCLCPP_INFO(get_logger(), "[Reliability]: Volatile");
            }else if(durability == "system_default"){
                qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT);
                RCLCPP_INFO(get_logger(), "[Reliability]: System Default");
            }else{
                RCLCPP_ERROR(get_logger(), "Selected Reliability QoS doesn't exists");
            }


            pub_ = create_publisher<std_msgs::msg::String>("chatter", qos_profile_pub_);
            timer_ = create_wall_timer(1s, std::bind(&SimpleQoSPublisher::TimerCallback, this));

            RCLCPP_INFO(get_logger(), "publishing at 1 HZ");
        }

    private:
        unsigned int counter_;
        rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
        rclcpp::TimerBase::SharedPtr timer_;
        rclcpp::QoS qos_profile_pub_;

        void TimerCallback()
        {
            std_msgs::msg::String msg = std_msgs::msg::String();
            msg.data = std::to_string(counter_);
            pub_->publish(msg);
            counter_ += 1;
        }
};


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleQoSPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}